libs/consumption_metrics/src/lib.rs

```diff
@@ -0,0 +1,221 @@
+            value,
+        }
+    }
+}
+
+/// ProxyConsumptionMetric struct that defines the format for one
+/// proxy metric entry
+/// i.e.
+///
+/// ```json
+/// {
+/// "metric": "remote_storage_size",
+/// "type": "incremental",
+/// "endpoint_id": "5d07d9ce9237c4cd845ea7918c0afa7d",
+/// "direction": "rx",
+/// "start_time": "2022-12-28T11:07:19.317310284Z",
+/// "stop_time": "2022-12-28T11:07:29.317310284Z",
+/// "idempotency_key": "2022-12-28 11:07:19.317310324 UTC-1-4019",
+/// "value": 12345454
+/// }
+/// ```
+#[serde_as]
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct ProxyConsumptionMetric {
+    #[serde_as(as = "DisplayFromStr")]
+    pub metric: ConsumptionMetricKind,
+    #[serde(rename = "type")]
+    pub metric_type: &'static str,
+    pub endpoint_id: String,
+    pub direction: String,
+    pub start_time: DateTime<Utc>,
+    pub stop_time: DateTime<Utc>,
+    pub idempotency_key: String,
+    // value is a delta against previously send metric
+    pub value: u64,
+}
+
+impl ProxyConsumptionMetric {
+    pub fn new_incremental(
+        metric: ConsumptionMetricKind,
+        endpoint_id: String,
+        direction: String,
+        value: u64,
+        node_id: String,
+        start_time: DateTime<Utc>,
+        stop_time: DateTime<Utc>,
+    ) -> Self {
+        Self {
+            metric,
+            metric_type: "incremental",
+            endpoint_id,
+            direction,
+            start_time,
+            stop_time,
+            // key that allows metric collector to distinguish unique events
+            idempotency_key: format!(
+                "{}-{}-{:04}",
+                Utc::now(),
+                node_id,
+                rand::thread_rng().gen_range(0..=9999)
+            ),
+            value,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Ord, PartialOrd, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum ConsumptionMetricKind {
+    /// Amount of WAL produced , by a timeline, i.e. last_record_lsn
+    /// This is an absolute, per-timeline metric.
+    WrittenSize,
+    /// Size of all tenant branches including WAL
+    /// This is an absolute, per-tenant metric.
+    /// This is the same metric that tenant/tenant_id/size endpoint returns.
+    SyntheticStorageSize,
+    /// Size of all the layer files in the tenant's directory on disk on the pageserver.
+    /// This is an absolute, per-tenant metric.
+    /// See also prometheus metric RESIDENT_PHYSICAL_SIZE.
+    ResidentSize,
+    /// Size of the remote storage (S3) directory.
+    /// This is an absolute, per-tenant metric.
+    RemoteStorageSize,
+    /// Logical size of the data in the timeline
+    /// This is an absolute, per-timeline metric
+    TimelineLogicalSize,
+    /// Number of bytes proxied
+    /// This is an incremental per endpoint proxy metric
+    /// see also NUM_BYTES_PROXIED_COUNTER
+    ProxyIOBytesPerClient,
+}
```

Comment:
I still don't quite get why we need to define this in `consumption_metrics`. This library is a dependency and as such can't be responsible for enumerating all possible metrics of its users. Besides, that'd mean that e.g. `ProxyIOBytesPerClient` is available in pageserver, which doesn't make sense.

Instead, I propose do define metric names in a following manner:

```rust
// proxy/src/proxy/metrics.rs
const PROXY_IO_BYTES_PER_CLIENT: &str = "...";
...
```

Then it'd be possible to reuse those names everywhere in the proxy (the same logic applies elsewhere).

In `PageserverConsumptionMetric` and `ProxyConsumptionMetric` we'd have

```rust
// because the name should be a compile-time constant
pub metric: &'static str,
```

Furthermore, `PageserverConsumptionMetric` & `ProxyConsumptionMetric` look like twins. It should be possible to unite them and even extract the common denominator. Here's how I would do that:

```rust
use chrono::{DateTime, Utc};
use serde::Serialize;

fn main() {
    // This could be literally anything as long as it implements Serialize.
    // I'd put contextually-dependedent things here (e.g. things which make
    // sense for the proxy but not pageserver).
    //
    // I've noticed that in the cloud OpenAPI spec absolute events use
    // tenant_id while incremental events use endpoint_id, which to me
    // doesn't make sense. However, if that's how it should truly work,
    // we could move them to EventType.
    //
    // If there are any fields that I forgot to describe, we should
    // embed them into Event.
    #[derive(Serialize)]
    struct Ids {
        tenant_id: String,
        timeline_id: String,
    }

    let event = Event {
        kind: EventType::Incremental {
            start_time: Utc::now(),
            stop_time: Utc::now(),
        },
        metric: "metric name goes here",
        idempotency_key: "...".into(),
        value: 100500,
        extra: Ids {
            tenant_id: "foo".into(),
            timeline_id: "bar".into(),
        },
    };

    println!("{}", serde_json::to_string(&event).unwrap());
}

#[derive(Serialize, Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
#[serde(tag = "type")]
pub enum EventType {
    #[serde(rename = "absolute")]
    Absolute {
        time: DateTime<Utc>,
    },
    #[serde(rename = "incremental")]
    Incremental {
        start_time: DateTime<Utc>,
        stop_time: DateTime<Utc>,
    },
}

#[derive(Serialize, Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
pub struct Event<Extra> {
    #[serde(flatten)]
    pub kind: EventType,

    pub metric: &'static str,
    pub idempotency_key: String,
    pub value: u64,

    #[serde(flatten)]
    pub extra: Extra,
}
```

```
$ cargo run
   Compiling jsons v0.1.0 (/tmp/jsons)
    Finished dev [unoptimized + debuginfo] target(s) in 0.48s
     Running `target/debug/jsons`
{"type":"incremental","start_time":"2023-01-12T12:39:06.596256607Z","stop_time":"2023-01-12T12:39:06.596262568Z","metric":"metric name goes here","idempotency_key":"...","value":100500,"tenant_id":"foo","timeline_id":"bar"}
```

One of the benefits is that it closely mirrors the OpenAPI spec and is substantially easier to comprehend.

Reply:
Thank you for the suggestion!
I pushed the refactoring